modules/gateway/peers.go

```diff
@@ -225,6 +222,61 @@ func (g *Gateway) threadedAcceptConn(conn net.Conn) {
 	g.log.Printf("INFO: accepted connection from new peer %v (v%v)", addr, remoteVersion)
 }
 
+// acceptableVersion returns an error if the version is unacceptable.
+func acceptableVersion(version string) error {
+	if !build.IsVersion(version) {
+		return invalidVersionError(version)
+	}
+	if build.VersionCmp(version, minAcceptableVersion) < 0 {
+		return insufficientVersionError(version)
+	}
+	return nil
+}
+
+// connectVersionHandshake performs the version handshake and should be called
+// on the side making the connection request. The remote version is only
+// returned if err == nil.
+func connectVersionHandshake(conn net.Conn, version string) (remoteVersion string, err error) {
+	// Send our version.
+	if err := encoding.WriteObject(conn, version); err != nil {
+		return "", fmt.Errorf("failed to write version: %v", err)
+	}
+	// Read remote version.
+	if err := encoding.ReadObject(conn, &remoteVersion, maxAddrLength); err != nil {
+		return "", fmt.Errorf("failed to read remote version: %v", err)
+	}
+	// Check that their version is acceptable.
+	if remoteVersion == "reject" {
+		return "", errPeerRejectedConn
+	}
+	if err := acceptableVersion(remoteVersion); err != nil {
+		return "", err
+	}
+	return remoteVersion, nil
+}
+
+// acceptConnVersionHandshake performs the version handshake and should be
+// called on the side accepting a connection request. The remote version is
+// only returned in err == nil.
+func acceptConnVersionHandshake(conn net.Conn, version string) (remoteVersion string, err error) {
+	// Read remote version.
+	if err := encoding.ReadObject(conn, &remoteVersion, maxAddrLength); err != nil {
+		return "", fmt.Errorf("failed to read remote version: %v", err)
+	}
+	// Check that their version is acceptable.
+	if err := acceptableVersion(remoteVersion); err != nil {
+		if err := encoding.WriteObject(conn, "reject"); err != nil {
+			return "", fmt.Errorf("failed to write reject: %v", err)
+		}
+		return "", err
+	}
+	// Send our version.
+	if err := encoding.WriteObject(conn, version); err != nil {
+		return "", fmt.Errorf("failed to write version: %v", err)
+	}
+	return remoteVersion, nil
+}
```

Comment:
These three functions look pretty good to me. I wish there was some way we could escape needing to pass in the version string as an argument, but it doesn't seem easily avoidable.